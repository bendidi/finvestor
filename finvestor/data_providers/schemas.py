import logging
import typing as tp
from collections.abc import Sequence
from datetime import datetime

import pandas as pd
import pytz
from pydantic import BaseModel, Field, validator

logger = logging.getLogger(__name__)

ValidInterval = tp.Literal[
    "1m", "2m", "5m", "15m", "30m", "1h", "1d", "5d", "1mo", "3mo"
]
AutoValidInterval = tp.Union[tp.Literal["auto"], ValidInterval]
ValidPeriod = tp.Literal[
    "1d", "5d", "1mo", "3mo", "6mo", "1y", "2y", "5y", "10y", "ytd"
]

VALID_INTERVALS: tp.Tuple[ValidInterval, ...] = tp.get_args(ValidInterval)


class Asset(BaseModel):
    ticker: str
    name: str
    type: str
    currency: str
    exchange: str
    exchange_timezone: str
    market: str
    isin: tp.Optional[str]
    country: tp.Optional[str]
    sector: tp.Optional[str]
    industry: tp.Optional[str]


class Bar(BaseModel):
    timestamp: datetime
    close: float
    high: float
    low: float
    open: float
    volume: float
    interval: ValidInterval


class Bars(BaseModel, Sequence):
    __root__: tp.List[Bar] = Field(..., min_items=1)

    def __getitem__(self, i):
        return self.__dict__.get("__root__")[i]

    def __len__(self):
        return len(self.__dict__.get("__root__"))

    def df(self) -> pd.DataFrame:
        bars_df = pd.DataFrame(self.dict().get("__root__"))
        bars_df["timestamp"] = pd.to_datetime(bars_df["timestamp"])
        bars_df = bars_df.set_index("timestamp")
        bars_df = bars_df.dropna(
            subset=["close", "open", "high", "low", "volume"], how="all"
        )
        return bars_df


class BaseBarsRequestParams(BaseModel):
    start: tp.Optional[datetime]
    end: tp.Optional[datetime]
    period: tp.Optional[ValidPeriod]
    interval: AutoValidInterval = "auto"

    @validator("period")
    def check_duplicate(cls, period: ValidPeriod, values) -> ValidPeriod:
        start = values.get("start")
        if start is None and period is None:
            raise ValueError("Missing values: 'period' or 'start'.")
        if start is not None and period is not None:
            raise ValueError("Duplicate values: 'period' and 'start'.")
        return period

    @validator("end")
    def check_invalid_end_value(
        cls, end: tp.Optional[datetime], values
    ) -> tp.Optional[datetime]:
        start = values.get("start")
        if end is not None and start is None:
            raise ValueError("Missing value: 'start'.")
        if start is not None and end is not None and end < start:
            raise ValueError("Invalid value: 'end' < 'start'.")
        return end

    @validator("start", "end")
    def check_tz_utc(cls, value: tp.Optional[datetime]) -> tp.Optional[datetime]:
        if value is None:
            return value
        if value.tzinfo is None or value.tzinfo.utcoffset(value) is None:
            raise ValueError(f"Missing timezone info: {value}")
        if value.tzinfo == pytz.UTC:
            return value
        return value.astimezone(pytz.UTC)

    def params(self, **kwargs: tp.Any) -> tp.Dict[str, tp.Any]:
        data = self.dict(exclude_none=True)
        data.update(kwargs)
        return data
