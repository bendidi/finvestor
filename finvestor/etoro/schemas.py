from datetime import datetime

import pandas as pd
from pydantic import BaseModel, Field, validator


class EtoroAccountSummary(BaseModel):
    name: str = Field(..., alias="Name")
    username: str = Field(..., alias="Username")
    currency: str = Field(..., alias="Currency")
    created_at: datetime = Field(..., alias="Date Created")
    start_date: datetime = Field(..., alias="Start Date")
    end_date: datetime = Field(..., alias="End Date")

    @validator("created_at", "start_date", "end_date", pre=True)
    def parse_dates(cls, value):
        return datetime.strptime(value, "%d/%m/%Y %H:%M:%S")

    initial_realised_equity: float = Field(..., alias="Beginning Realized Equity")
    initial_unrealised_equity: float = Field(..., alias="Beginning Unrealized Equity")
    deposits: float = Field(..., alias="Deposits")
    refunds: float = Field(..., alias="Refunds")
    credits: float = Field(..., alias="Credits")
    adjustments: float = Field(..., alias="Adjustments")
    withdrawals: float = Field(..., alias="Withdrawals")
    realised_profit_loss: float = Field(
        ..., alias="Profit or Loss (Closed positions only)"
    )
    rollover_fees: float = Field(..., alias="Rollover Fees")
    withdrawal_fees: float = Field(..., alias="Withdrawal Fees")
    final_realised_equity: float = Field(..., alias="Ending Realized Equity")
    final_unrealised_equity: float = Field(..., alias="Ending Unrealized Equity")


class EtoroFinancialSummary(BaseModel):
    cfd_profit_loss: float = Field(..., alias="CFDs (Profit or Loss)")
    crypto_profit_loss: float = Field(..., alias="Crypto (Profit or Loss)")
    stocks_profit_loss: float = Field(..., alias="Stocks (Profit or Loss)")
    etf_profit_loss: float = Field(..., alias="ETFs (Profit or Loss)")
    stock_dividends_profit: float = Field(..., alias="Stock Dividends (Profit)", ge=0)
    cfd_dividends_profit_loss: float = Field(
        ..., alias="CFD Dividends (Profit or Loss)"
    )
    refunds: float = Field(..., alias="Income from Refunds")
    cfd_commissions: float = Field(..., alias="Commissions (spread) on CFDs")
    crypto_commissions: float = Field(..., alias="Commissions (spread) on Crypto")
    etf_commissions: float = Field(..., alias="Commissions (spread) on ETFs")
    fees: float = Field(..., alias="Fees")


class EtoroAccountStatement(BaseModel):
    account_summary: EtoroAccountSummary
    financial_summary: EtoroFinancialSummary
    closed_positions: pd.DataFrame
    account_activity: pd.DataFrame

    class Config:
        arbitrary_types_allowed = True
