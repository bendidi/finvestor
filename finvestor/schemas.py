from datetime import datetime
from typing import Optional, Literal
from pydantic import BaseModel
from pydantic.fields import Field


class Asset(BaseModel):
    symbol: str
    currency: str = "USD"
    name: Optional[str]
    category: Optional[
        Literal["stock", "crypto", "ETF", "currency", "commodity", "cash"]
    ]
    industry: Optional[str]
    exchange: Optional[str]


class Transaction(BaseModel):
    asset: Asset
    type: Literal["BUY", "SELL", "DEPOSIT", "WITHDRAW"]
    open_date: datetime
    close_date: Optional[datetime]
    open_rate: float
    close_rate: Optional[float]
    units: float
    leverage: float = Field(default=1, ge=1)
    spread: float = 0
    fees: float = 0
    profit: Optional[float]
