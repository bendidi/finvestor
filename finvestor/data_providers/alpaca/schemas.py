import typing as tp
from datetime import datetime

import pytz
from pydantic import Field

from finvestor.core.timeframe import parse_duration
from finvestor.data_providers.alpaca.settings import alpaca_settings
from finvestor.data_providers.alpaca.utils import convert_interval_to_alpaca_timeframe
from finvestor.data_providers.schemas import BaseBarsRequestParams


class APCABarsRequestParams(BaseBarsRequestParams):

    limit: tp.Optional[int] = Field(None, ge=1, le=10000)
    adjustment: tp.Literal[None, "raw", "split", "dividend", "all"] = None

    def params(
        self, end_default: tp.Literal["now", "interval"] = "now", **kwargs: tp.Any
    ) -> tp.Dict[str, tp.Any]:
        data = super().params(**kwargs)

        # set timeframe
        data["timeframe"] = convert_interval_to_alpaca_timeframe(data.pop("interval"))

        # if we have start but not end
        if "start" in data:
            if "end" not in data:
                data["end"] = (
                    datetime.now(tz=pytz.UTC) - alpaca_settings.HISTORICAL_DATA_DELAY
                    if end_default == "now"
                    # we set it to start + interval*2 to make sure we get enough data
                    else data["start"] + parse_duration(data["timeframe"]) * 2
                )

        # convert period to start-end
        # TODO: doesn't take into account days where the market is not Open
        # (holidays, weekends, ..)
        # maybe have a look at: https://github.com/rsheftel/pandas_market_calendars
        if "period" in data:
            period = data.pop("period")
            data["end"] = (
                datetime.now(tz=pytz.UTC) - alpaca_settings.HISTORICAL_DATA_DELAY
            )
            data["start"] = data["end"] - parse_duration(period)

        data["start"] = data["start"].isoformat()
        data["end"] = data["end"].isoformat()
        return data
