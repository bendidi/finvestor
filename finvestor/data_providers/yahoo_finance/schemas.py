import typing as tp
from datetime import datetime

import pytz
from pydantic import BaseModel

from finvestor.core.timeframe import parse_duration
from finvestor.data_providers.schemas import BaseBarsRequestParams
from finvestor.data_providers.yahoo_finance.utils import get_smallest_valid_interval


class YFBarsRequestParams(BaseBarsRequestParams):
    include_prepost: tp.Optional[bool] = None
    events: tp.Literal[None, "div", "split", "div,splits"] = None

    def params(
        self, end_default: tp.Literal["now", "interval"] = "now", **kwargs: tp.Any
    ) -> tp.Dict[str, tp.Any]:
        data = super().params(**kwargs)
        if data["interval"] == "auto":
            data["interval"] = get_smallest_valid_interval(**data)
        if "include_prepost" in data:
            data["includePrePost"] = data.pop("include_prepost")
        if "period" in data:
            data["range"] = data.pop("period")

        if "start" in data:
            if "end" not in data:
                data["end"] = (
                    datetime.now(tz=pytz.UTC)
                    if end_default == "now"
                    # we set it to start + interval*2 to make sure we get enough data
                    else data["start"] + parse_duration(data["interval"]) * 2
                )
            data["period1"] = int(datetime.timestamp(data.pop("start")))
            data["period2"] = int(datetime.timestamp(data.pop("end")))
        return data


class YfOpenHighLowClose(BaseModel):
    timestamp: tp.List[datetime]
    volume: tp.List[tp.Optional[float]]
    open: tp.List[tp.Optional[float]]
    high: tp.List[tp.Optional[float]]
    low: tp.List[tp.Optional[float]]
    close: tp.List[tp.Optional[float]]
