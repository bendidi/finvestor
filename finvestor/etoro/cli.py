from pathlib import Path
import os
import typer

from finvestor.etoro.io import load_etoro_account_statement
from finvestor.etoro.portfolio import get_deposits_breakdown

app = typer.Typer(help="Load and process an etoro account statement.")


@app.callback()
def main(
    filepath: Path = typer.Option(
        ...,
        "-f",
        "--file",
        exists=True,
        file_okay=True,
        dir_okay=False,
        writable=False,
        readable=True,
        help="Path to etoro_account_statement.xlsx",
    )
):
    """
    Load and process an etoro account statement.
    """
    typer.secho(
        f"Loading Etoro account statement '{os.path.basename(filepath)}'...",
        fg=typer.colors.BRIGHT_GREEN,
    )
    statement = load_etoro_account_statement(filepath)
    typer.secho(
        f"Loaded Etoro account statement of user '{statement.account_summary.name}' "
        f"from '{statement.account_summary.start_date}' "
        f"to '{statement.account_summary.end_date}'",
        fg=typer.colors.BRIGHT_GREEN,
    )
    get_deposits_breakdown(statement)
