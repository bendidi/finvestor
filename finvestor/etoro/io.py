import pandas as pd
from pathlib import Path

from typing import Tuple, Union

from finvestor.etoro.schemas import (
    EtoroAccountStatement,
    EtoroAccountSummary,
    EtoroFinancialSummary,
)


def load_etoro_account_statement(filepath: Union[str, Path]) -> EtoroAccountStatement:
    """Load an etoro account statement xlsx file.

    Args:
        filepath (str): path to xlsx file.
            can also be: bytes, ExcelFile, xlrd.Book, path object, or file-like object

    Returns:
        EtoroAccountStatement
    """
    all_sheets = pd.read_excel(filepath, sheet_name=None)

    closed_positions_df = pre_process_closed_positions_df(
        all_sheets["Closed Positions"]
    )
    (
        fees_df,
        deposits_df,
        account_activity_open_positions_df,
        account_activity_closed_positions_df,
    ) = pre_process_account_activity_df(all_sheets["Account Activity"])

    transactions_df = aggregate_transactions_df(
        closed_positions_df,
        account_activity_open_positions_df,
        account_activity_closed_positions_df,
    )
    return EtoroAccountStatement(
        account_summary=parse_account_summary(all_sheets["Account Summary"]),
        financial_summary=parse_financial_summary(all_sheets["Financial Summary"]),
        transactions_df=transactions_df,
        fees_df=fees_df,
        deposits_df=deposits_df,
    )


def aggregate_transactions_df(
    closed_positions_df: pd.DataFrame,
    account_activity_open_positions_df: pd.DataFrame,
    account_activity_closed_positions_df: pd.DataFrame,
) -> pd.DataFrame:
    """Aggregate all transactions found in an etoro account statement.

    Args:
        closed_positions_df: pre-processes closed positions sheets found in an
            etoro account statement
        account_activity_open_positions_df: Filtered open positions found in
            the account activit sheet of an etoro account statement
        account_activity_closed_positions_df: Filtered closed positions found in
            the account activit sheet of an etoro account statement

    Returns:
        pd.DataFrame: All transaction aggregated in one single dataframe
    """
    assert len(closed_positions_df) == len(
        account_activity_closed_positions_df
    ), "Invalid or corrupt data."

    df = closed_positions_df.merge(
        account_activity_closed_positions_df, on="position_id"
    )
    df = df.drop(columns=["amount", "date"])

    df = account_activity_open_positions_df.drop(columns=["date"]).merge(
        df.drop(columns=["invested", "realized_equity"]),
        on=("position_id", "details"),
        how="left",
        suffixes=("_open", "_close"),
    )
    return df


def parse_account_summary(df: pd.DataFrame) -> EtoroAccountSummary:
    """Parse the account summary object from pandas dataframe.

    Args:
        df: pandas dataframe of 'Account Summary' sheet loaded from etoro account
            statement xlsx

    Returns:
        EtoroAccountSummary
    """
    # apply some preprocessing
    df = (
        df.melt(id_vars=["Details"])
        .drop("variable", axis=1)
        .dropna()
        .drop(df.index[[0, 8, 20]])
    )
    return EtoroAccountSummary(**dict(zip(df.Details, df.value)))


def parse_financial_summary(df: pd.DataFrame) -> EtoroFinancialSummary:
    """Load the financial summary object from pandas dataframe.

    Args:
        df: pandas dataframe of 'Financial Summary' sheet loaded from etoro account
            statement xlsx

    Returns:
        EtoroFinancialSummary
    """
    # apply some preprocessing
    df = df.drop("Tax\nRate", axis=1).dropna()
    return EtoroFinancialSummary(**dict(zip(df.Name, df["Amount\nin USD"])))


def pre_process_account_activity_df(
    df: pd.DataFrame,
) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]:
    """Pre-process the account activity sheet from etoro account statement.

    -> Drop the column 'NWA', no information about it yet.
    -> Transform column names to lower string and replace spaces with '_'
    -> Split dataframe into:
        - fees dataframe
        - deposits dataframe
        - withdrawals dataframe
        - open positions dataframe
        - closed positions dataframe
    -> Convert all values of column 'type' to upper case (Buy -> BUY, Sell -> SELL)

    Args:
        df: The account activity dataframe.

    Returns:
        Tuple[fees_df, deposits_df, open_df, closed_df]
    """

    # drop NWA clumn: don't what it means :)
    df = df.drop(columns=["NWA"])
    df.columns = df.columns.str.lower().str.replace(" ", "_")

    # extract all gees into a seprate dataframe
    fees_df = df[(df.type == "Adjustment") | (df.type == "Rollover Fee")]

    deposits_df = df[df.type == "Deposit"]
    deposits_df = deposits_df.drop(columns=["type", "position_id"])

    withdrawals_df = df[
        (df.type == "Withdraw Fee")
        | (df.type == "Withdraw Request")
        | (df.type == "Withdraw Fee Cancelled")
        | (df.type == "Withdraw Request Cancelled")
    ]
    withdrawals_df = withdrawals_df.drop(columns=["details", "position_id"])

    open_df = df[df.type == "Open Position"]
    open_df = open_df.drop(columns=["type", "realized_equity_change"])
    open_df = open_df.rename(columns={"amount": "invested"})

    closed_df = df[df.type == "Profit/Loss of Trade"]
    closed_df = closed_df.drop(columns=["type"])
    return fees_df, deposits_df, open_df, closed_df


def pre_process_closed_positions_df(df: pd.DataFrame) -> pd.DataFrame:
    """Pre-process the closed positions sheet from etoro account statement.

    -> Split the 'Action' column into 'type' (Buy or Sell) and 'name'
        (full name of the company)
    -> Drop the columns: 'Copied From', 'Type', 'Notes', 'Action'
    -> Transform column names to lower string and replace spaces with '_'
    -> rename some columns:
        - stop_lose_rate: stop_loss_rate
        - amount: invested
        - isin: ISIN
    -> Convert all values of column 'type' to upper case (Buy -> BUY, Sell -> SELL)

    Args:
        df: The closed positions sheet as a pandas dataframe.

    Returns:
        pd.DataFrame
    """
    # split string by first space, to get type of transacation (BUY/SELL) and
    # company name
    df[["type", "name"]] = df["Action"].str.split(" ", n=1, expand=True)

    # drop unnecessary columns
    df = df.drop(columns=["Copied From", "Type", "Notes", "Action"])

    # convert column names to lower case and replace spaces with '_'
    df.columns = df.columns.str.lower().str.replace(" ", "_")

    # rename some columns to be clearer or fix typos
    df = df.rename(
        columns={
            "stop_lose_rate": "stop_loss_rate",
            "amount": "invested",
            "isin": "ISIN",
        }
    )

    # convert column 'type' to uper case (Buy -> BUY, Sell -> SELL)
    df["type"] = df["type"].str.upper()

    return df
