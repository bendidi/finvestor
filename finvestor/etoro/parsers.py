import pandas as pd

from typing import Tuple
from finvestor.etoro.schemas import EtoroAccountSummary, EtoroFinancialSummary


def aggregate_transactions_df(
    closed_positions_df: pd.DataFrame,
    account_activity_open_positions_df: pd.DataFrame,
    account_activity_closed_positions_df: pd.DataFrame,
) -> pd.DataFrame:
    """Aggregate all transactions found in an etoro account statement.

    Args:
        closed_positions_df: pre-processes closed positions sheets found in an
            etoro account statement
        account_activity_open_positions_df: Filtered open positions found in
            the account activit sheet of an etoro account statement
        account_activity_closed_positions_df: Filtered closed positions found in
            the account activit sheet of an etoro account statement

    Returns:
        pd.DataFrame: All transaction aggregated in one single dataframe
    """
    assert len(closed_positions_df) == len(
        account_activity_closed_positions_df
    ), "Invalid or corrupt data."

    df = closed_positions_df.merge(
        account_activity_closed_positions_df, on="position_id"
    )
    df = df.drop(columns=["amount", "date"], errors="ignore")

    df = account_activity_open_positions_df.drop(
        columns=["date"], errors="ignore"
    ).merge(
        df.drop(columns=["invested", "realized_equity"], errors="ignore"),
        on=("position_id", "details"),
        how="left",
        suffixes=("_open", "_close"),
    )
    return df


def parse_account_summary(df: pd.DataFrame) -> EtoroAccountSummary:
    """Parse the account summary object from pandas dataframe.

    Args:
        df: pandas dataframe of 'Account Summary' sheet loaded from etoro account
            statement xlsx

    Returns:
        EtoroAccountSummary
    """
    # apply some preprocessing
    df = (
        df.melt(id_vars=["Details"])
        .drop("variable", axis=1)
        .dropna()
        .drop(df.index[[0, 8, 20]])
    )
    return EtoroAccountSummary(**dict(zip(df.Details, df.value)))


def parse_financial_summary(df: pd.DataFrame) -> EtoroFinancialSummary:
    """Load the financial summary object from pandas dataframe.

    Args:
        df: pandas dataframe of 'Financial Summary' sheet loaded from etoro account
            statement xlsx

    Returns:
        EtoroFinancialSummary
    """
    # apply some preprocessing
    df = df.drop("Tax\nRate", axis=1).dropna()
    return EtoroFinancialSummary(**dict(zip(df.Name, df["Amount\nin USD"])))


def pre_process_account_activity_df(
    df: pd.DataFrame,
) -> Tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame, pd.DataFrame]:
    """Pre-process the account activity sheet from etoro account statement.

    -> Drop the column 'NWA', no information about it yet.
    -> Transform column names to lower string and replace spaces with '_'
    -> Split dataframe into:
        - fees dataframe
        - deposits dataframe
        - withdrawals dataframe
        - open positions dataframe
        - closed positions dataframe
    -> Convert all values of column 'type' to upper case (Buy -> BUY, Sell -> SELL)

    Args:
        df: The account activity dataframe.

    Returns:
        Tuple[fees_df, deposits_df, open_df, closed_df]
    """

    # drop NWA clumn: don't what it means :)
    df = df.drop(columns=["NWA"], errors="ignore")
    df.columns = df.columns.str.lower().str.replace(" ", "_")

    # extract all gees into a seprate dataframe
    fees_df = df[(df.type == "Adjustment") | (df.type == "Rollover Fee")]

    deposits_df = df[df.type == "Deposit"]
    deposits_df = deposits_df.drop(columns=["type", "position_id"], errors="ignore")

    withdrawals_df = df[
        (df.type == "Withdraw Fee")
        | (df.type == "Withdraw Request")
        | (df.type == "Withdraw Fee Cancelled")
        | (df.type == "Withdraw Request Cancelled")
    ]
    withdrawals_df = withdrawals_df.drop(
        columns=["details", "position_id"], errors="ignore"
    )

    open_df = df[df.type == "Open Position"]
    open_df = open_df.drop(columns=["type", "realized_equity_change"], errors="ignore")
    open_df = open_df.rename(columns={"amount": "invested"})

    closed_df = df[df.type == "Profit/Loss of Trade"]
    closed_df = closed_df.drop(columns=["type"], errors="ignore")
    return fees_df, deposits_df, open_df, closed_df


def pre_process_closed_positions_df(df: pd.DataFrame) -> pd.DataFrame:
    """Pre-process the closed positions sheet from etoro account statement.

    -> Split the 'Action' column into 'type' (Buy or Sell) and 'name'
        (full name of the company)
    -> Drop the columns: 'Copied From', 'Type', 'Notes', 'Action'
    -> Transform column names to lower string and replace spaces with '_'
    -> rename some columns:
        - stop_lose_rate: stop_loss_rate
        - amount: invested
        - isin: ISIN
    -> Convert all values of column 'type' to upper case (Buy -> BUY, Sell -> SELL)

    Args:
        df: The closed positions sheet as a pandas dataframe.

    Returns:
        pd.DataFrame
    """
    # split string by first space, to get type of transacation (BUY/SELL) and
    # company name
    df[["type", "name"]] = df["Action"].str.split(" ", n=1, expand=True)

    # drop unnecessary columns
    df = df.drop(columns=["Copied From", "Type", "Notes", "Action"], errors="ignore")

    # convert column names to lower case and replace spaces with '_'
    df.columns = df.columns.str.lower().str.replace(" ", "_")

    # rename some columns to be clearer or fix typos
    df = df.rename(
        columns={
            "stop_lose_rate": "stop_loss_rate",
            "amount": "invested",
            "isin": "ISIN",
        }
    )

    # convert column 'type' to uper case (Buy -> BUY, Sell -> SELL)
    df["type"] = df["type"].str.upper()

    return df
