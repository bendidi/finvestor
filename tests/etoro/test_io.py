from finvestor.etoro.io import (
    parse_account_summary,
    parse_financial_summary,
    pre_process_closed_positions_df,
)
from finvestor.etoro.schemas import EtoroAccountSummary, EtoroFinancialSummary


def test_parse_account_summary(snapshot, etoro_account_summary_df):
    parsed = parse_account_summary(etoro_account_summary_df)
    assert isinstance(parsed, EtoroAccountSummary)
    assert parsed == snapshot


def test_parse_financial_summary(snapshot, etoro_financial_summary_df):
    parsed = parse_financial_summary(etoro_financial_summary_df)
    assert isinstance(parsed, EtoroFinancialSummary)
    assert parsed == snapshot


def test_pre_process_closed_positions_df(snapshot, etoro_closed_positions_df):
    df = pre_process_closed_positions_df(etoro_closed_positions_df)
    assert df.to_dict() == snapshot
