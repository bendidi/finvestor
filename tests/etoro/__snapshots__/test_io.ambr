# name: test_parse_account_summary
  <class 'dict'> {
    'adjustments': 9.87,
    'created_at': datetime.datetime(1999, 1, 1, 10, 28, 13),
    'credits': 0.0,
    'currency': 'USD',
    'deposits': 123456.789,
    'end_date': datetime.datetime(1999, 1, 1, 23, 59, 59),
    'final_realised_equity': 1101079.94,
    'final_unrealised_equity': 1103079.94,
    'initial_realised_equity': 0.0,
    'initial_unrealised_equity': 0.0,
    'name': 'John Doe',
    'realised_profit_loss': 987654.321,
    'refunds': 0.0,
    'rollover_fees': 1.17,
    'start_date': datetime.datetime(1998, 1, 1, 0, 0),
    'username': 'apes_together_strong',
    'withdrawal_fees': 30.0,
    'withdrawals': 10000.0,
  }
---
# name: test_parse_financial_summary
  <class 'dict'> {
    'cfd_commissions': 60.87,
    'cfd_dividends_profit_loss': 3.45,
    'cfd_profit_loss': 7654.0,
    'crypto_commissions': 2.01,
    'crypto_profit_loss': 8.8,
    'etf_commissions': 0.0,
    'etf_profit_loss': 0.0,
    'fees': 8.77,
    'refunds': 2.6,
    'stock_dividends_profit': 6.49,
    'stocks_profit_loss': 98000.321,
  }
---
# name: test_pre_process_closed_positions_df
  <class 'dict'> {
    'ISIN': <class 'dict'> {
      0: 'US215621452',
      1: 'CH015236589',
    },
    'close_date': <class 'dict'> {
      0: '12/06/2000 16:33:45',
      1: '07/02/2002 15:12:15',
    },
    'close_rate': <class 'dict'> {
      0: 4.02,
      1: 215.32,
    },
    'invested': <class 'dict'> {
      0: 1500.0,
      1: 750.0,
    },
    'leverage': <class 'dict'> {
      0: 1,
      1: 5,
    },
    'name': <class 'dict'> {
      0: 'Awesome-stock Inc.',
      1: 'Bad company',
    },
    'open_date': <class 'dict'> {
      0: '01/01/2000 09:30:07',
      1: '01/01/2000 10:10:15',
    },
    'open_rate': <class 'dict'> {
      0: 2.99,
      1: 115.0,
    },
    'position_id': <class 'dict'> {
      0: 1,
      1: 2,
    },
    'profit': <class 'dict'> {
      0: 512.65,
      1: -654.21,
    },
    'rollover_fees_and_dividends': <class 'dict'> {
      0: 0.0,
      1: -2.03,
    },
    'spread': <class 'dict'> {
      0: 0.0,
      1: 10.43,
    },
    'stop_loss_rate': <class 'dict'> {
      0: 1,
      1: 300,
    },
    'take_profit_rate': <class 'dict'> {
      0: 10,
      1: 100,
    },
    'type': <class 'dict'> {
      0: 'BUY',
      1: 'SELL',
    },
    'units': <class 'dict'> {
      0: 500.12,
      1: 6.52146,
    },
  }
---
